### PR TITLE
Add edit/add/delete view

### LINKED ISSUE #1961: Allow editing views

### LINKED ISSUE #1962: Allow adding views

### LINKED ISSUE #1963: Allow removing views

### PR DESCRIPTION
**Pull request by bramkragten**
views
Closes: https://github.com/home-assistant/home-assistant-polymer/issues/1961 Closes: https://github.com/home-assistant/home-assistant-polymer/issues/1962 Closes: https://github.com/home-assistant/home-assistant-polymer/issues/1963

### CODE REVIEW

**Comment by pepeEL** [comment]
Please also add strings to translation to we can start working on translating.

File: src/panels/lovelace/hui-root.js [hunk 1]
```diff
@@ -295,6 +323,50 @@ class HUIRoot extends NavigateMixin(EventsMixin(PolymerElement)) {
     this._selectView(this._curView);
   }
 
+  _editView() {
+    const viewConfig = this.config.views[this._curView];
+    if (viewConfig.cards) {
+      delete viewConfig.cards;
```
Comment (balloob) [inline src/panels/lovelace/hui-root.js]: Why delete this? Shouldn't we just make a copy and mutate that?

```ts
const { cards, badges, ...viewConfig } = this.config.views[this._curView];
```

File: src/panels/lovelace/editor/hui-edit-view.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+  }
+
+  private _valueChanged(ev: Event): void {
+    if (!this._config || !this.hass) {
+      return;
+    }
+
+    const target = ev.currentTarget! as EditorTarget;
+
+    if (
```
Comment (balloob) [inline src/panels/lovelace/editor/hui-edit-view.ts]:
```ts
if (this[`_${target.configValue})`] === target.value) { return; }
```

File: src/panels/lovelace/editor/hui-edit-view.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        await updateViewConfig(
+          this.hass!,
+          this.viewConfig!.id!,
+          this._config!,
+          "json"
+        );
+      }
+      fireEvent(this, "reload-lovelace");
+      this._closeDialog();
+      this._saveDone();
```
Comment (balloob) [inline src/panels/lovelace/editor/hui-edit-view.ts]: Lets just inline this function as the name is really confusing, especially combined with previous 2 lines.

File: src/panels/lovelace/editor/hui-dialog-edit-view.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        .add="${this._params.add}"
+        @reload-lovelace="${this._params.reloadLovelace}"
+        @cancel-edit-view="${this._cancel}"
+      >
+      </hui-edit-view>
+    `;
+  }
+
+  private _cancel() {
+    this._params = {
```
Comment (balloob) [inline src/panels/lovelace/editor/hui-dialog-edit-view.ts]: Why is this needed?

And it feels a little weird that `hui-edit-view` is closing it's dialog and then have to tell parent to clean up too.

File: src/panels/lovelace/editor/hui-dialog-edit-view.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+          @reload-lovelace="${this._params.reloadLovelace}"
+        ></hui-migrate-config>
+      `;
+    }
+    return html`
+      <hui-edit-view
+        .hass="${this.hass}"
+        .viewConfig="${this._params.viewConfig}"
+        .add="${this._params.add}"
+        @reload-lovelace="${this._params.reloadLovelace}"
```
Comment (balloob) [inline src/panels/lovelace/editor/hui-dialog-edit-view.ts]: We could pass this method as a prop instead of relying on events. Same with cancel edit.

Comment (balloob) [reply src/panels/lovelace/editor/hui-dialog-edit-view.ts]: We should use events when it's very high up in the chain and we don't know who exactly handles/fires it.

If it's a parent/child relationship, feel free to just pass methods down. Make sure you don't pass `this._cancel` though, as that won't bind the method correctly.

**Comment by bramkragten** [comment]
The translations are a breaking change :-(

File: src/panels/lovelace/editor/hui-dialog-edit-view.ts [hunk 6]
```diff
@@ -85,8 +85,7 @@ export class HuiDialogEditView extends LitElement {
         .hass="${this.hass}"
         .viewConfig="${this._params.viewConfig}"
         .add="${this._params.add}"
-        @reload-lovelace="${this._params.reloadLovelace}"
-        @cancel-edit-view="${this._cancel}"
```
Comment (balloob) [inline src/panels/lovelace/editor/hui-dialog-edit-view.ts]: Do you have to remove the registration of these events with typescript?

Comment (bramkragten) [reply src/panels/lovelace/editor/hui-dialog-edit-view.ts]: I removed `cancel-edit-view` the other one is still used

File: src/panels/lovelace/hui-root.js [hunk 7]
```diff
@@ -295,6 +323,44 @@ class HUIRoot extends NavigateMixin(EventsMixin(PolymerElement)) {
...
+        this.fire("config-refresh");
+      },
+    });
+  }
+
+  _addView() {
+    showEditViewDialog(this, {
+      add: true,
+      reloadLovelace: () => {
+        this.fire("config-refresh");
```
[the remark this answers is not in the document]
Comment (balloob) [reply src/panels/lovelace/hui-root.js]: could you add event detail to say "refresh and pick view" ?
